pkg/pillar/cas/cas.go

```diff
@@ -0,0 +1,89 @@
+	//Returns error if no blob is found for the given 'blobHash'.
+	//Arg 'blobHash' should be of format <algo>:<hash> (currently supporting only sha256:<hash>).
+	ReadBlob(blobHash string) (io.Reader, error)
+	//RemoveBlob: removes a blob which matches the given arg 'blobHash'.
+	//To keep this method idempotent, no error is returned if the given arg 'blobHash' does not match any blob.
+	//Arg 'blobHash' should be of format <algo>:<hash> (currently supporting only sha256:<hash>).
+	RemoveBlob(blobHash string) error
+	//Children: returns a list of child blob hashes if the given arg 'blobHash' belongs to a
+	// index or a manifest blob, else an empty list is returned.
+	//Format of returned blob hash list and arg 'blobHash' is <algo>:<hash> (currently supporting only sha256:<hash>)
+	Children(blobHash string) ([]string, error)
+
+	//Image APIs
+	//CreateImage: creates a reference which points to a blob with 'blobHash'. 'blobHash' must belong to a index blob
+	//Arg 'blobHash' should be of format <algo>:<hash> (currently supporting only sha256:<hash>).
+	//Returns error if no blob is found matching the given 'blobHash' or if the given 'blobHash' does not belong to an index.
+	CreateImage(reference, blobHash string) error
+	//GetImageHash: returns a blob hash of format <algo>:<hash> (currently supporting only sha256:<hash>) which the given 'reference' is pointing to.
+	// Returns error if the given 'reference' is not found.
+	GetImageHash(reference string) (string, error)
+	//ListImages: returns a list of references
+	ListImages() ([]string, error)
+	//RemoveImage: removes an reference along with its contents unless the contents are referred by another reference or snapshot.
+	//To keep this method idempotent, no error  is returned if the given 'reference' is not found.
+	RemoveImage(reference string) error
+	//ReplaceImage: replaces the blob hash to which the given 'reference' is pointing to with the given 'blobHash'.
+	//Returns error if the given 'reference' or a blob matching the given arg 'blobHash' is not found.
+	//Returns if the given 'blobHash' does not belong to an index.
+	//Arg 'blobHash' should be of format <algo>:<hash> (currently supporting only sha256:<hash>).
+	ReplaceImage(reference, blobHash string) error
+
+	//Snapshot APIs
+	//CreateSnapshotForImage: creates an snapshot with the given snapshotID for the given 'reference'
+	//Arg 'snapshotID' should be of format <algo>:<hash> (currently supporting only sha256:<hash>).
+	CreateSnapshotForImage(snapshotID, reference string) error
+	//MountSnapshot: mounts the snapshot on the given target path
+	//Arg 'snapshotID' should be of format <algo>:<hash> (currently supporting only sha256:<hash>).
+	MountSnapshot(snapshotID, targetPath string) error
+	//ListSnapshots: returns a list of snapshotIDs where each entry is of format <algo>:<hash> (currently supporting only sha256:<hash>).
+	ListSnapshots() ([]string, error)
```

Comment:
If we have two containers running using the identical contentTree/image, would we have shared shapshot or two different ones? If the latter then the sha isn't sufficient to identify the snapshot.

Reply:
> two containers running using the identical contentTree/image, would we have shared shapshot or two different ones?

Definitely two different ones, as it should be. The committed layers (snapshots of immutable content) can and should be the same. The active layer (where each container accesses) should be different. 

The consumer itself doesn't care about the lower layers, only that "here is my content tree, give me a snapshot I can work with." It should be able to do it multiple times, each time getting a distinct one (even if the lower layers are the same).

Reply:
The net of this is that we _will_ need to name snapshots (or at least, as they are exposed to consumers via the CAS API) something _other than_ `sha256:<hash>`. It just has no meaning that way.

Nice catch @eriknordmark 

I still think I would merge this in, and then immediately open one to fix it.

Reply:
OK